extensions-contrib/opentsdb-emitter/src/main/java/io/druid/emitter/opentsdb/OpentsdbSender.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.emitter.opentsdb;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.sun.jersey.api.client.Client;
+import com.sun.jersey.api.client.WebResource;
+import io.druid.java.util.common.logger.Logger;
+
+import javax.ws.rs.core.MediaType;
+import java.util.ArrayList;
+import java.util.List;
+
+public class OpentsdbSender
+{
+  /**
+   * @see <a href="http://opentsdb.net/docs/build/html/api_http/put.html">Opentsdb - /api/put</a>
+   */
+  private static final String PATH = "/api/put";
+  private static final Logger log = new Logger(OpentsdbSender.class);
+
+  private final int batchSize;
+  private final List<OpentsdbEvent> events;
+  private final Client client;
+  private final WebResource webResource;
+
+  public OpentsdbSender(String host, int port, int connectionTimeout, int readTimeout, int batchSize)
+  {
+    this.batchSize = batchSize;
+    events = new ArrayList<>(batchSize);
+
+    client = Client.create();
+    client.setConnectTimeout(connectionTimeout);
+    client.setReadTimeout(readTimeout);
+    webResource = client.resource("http://" + host + ":" + port + PATH);
+  }
+
+  public void send(OpentsdbEvent event)
+  {
+    events.add(event);
+    if (events.size() >= batchSize) {
```

Comment:
Maybe good to add a limit to the total number of events that can be queued (suppose the send takes a while because of network issues and lots of events queue up in the meantime, leading to high memory usage)

Reply:
How about using a BlockingQueue?

Reply:
A BlockingQueue would cause callers of emit() to block on the metrics event queue, I was thinking of having a configurable event limit, where events are discarded if events.size() exceeds that limit (and probably have sendEvents() send up to the batch size and not the entire event list)

Reply:
I got your point. I was thinking of a BlockingQueue maybe good if this emitter will be used in a multi-thread environment. But I am not sure about this. Is a emitter will be shared by multi threads ?

Reply:
For both single thread and multi thread contexts, I don't think adding a metrics event to an emitter should be a blocking call, I think dropping events if there's no capacity is the right approach.

Reply:
@QiuMM Ah, sorry, I think might have misunderstood how you were planning on using BlockingQueue. 

A BlockingQueue should be fine as long as you use offer() instead of add(), throwing away the events that couldn't be added due to capacity.

Reply:
Never mind. And I was going to refactor my code in the way just like what you said.